### PR TITLE
fix(durable): harden legacy reconciliation

### LINKED ISSUE #71: Add dual OpenClaw 7.1-2 and 8.1 SDK compatibility
## Goal

Add OpenClaw 2026.8.1 SDK compatibility while retaining support for the currently deployed OpenClaw 2026.7.1-2 release.

## Confirmed compatibility gaps

- The durable inbound receive helper moved from `createDurableInboundReceiveJournal` with keyed stores to `createDurableInboundReceiveJournalFromQueue` with the host channel ingress queue.
- `commands.useAccessGroups` was removed from `CommandsConfig`; access groups are enabled through the shared ingress path in 2026.8.1.
- `openclaw/plugin-sdk/outbound-media` ships runtime JavaScript but omits public TypeScript declarations in 2026.8.1.

## Compatibility constraints

- Keep the declared minimum host version at 2026.7.1-2 unless maintaining dual compatibility is proven impossible.
- Do not import removed or private OpenClaw internals.
- Do not require a live Gateway upgrade to develop or verify the change.
- Preserve authorization behavior and fail closed.
- Preserve durable receive semantics across 7.1-2 and 8.1, including completion, release, duplicate detection, and retention.
- Do not silently abandon pending 7.1-2 durable records; document or implement any required transition behavior.

## Acceptance criteria

- TypeScript builds against exact OpenClaw 2026.7.1-2 and 2026.8.1 packages.
- Unit and offline smoke tests pass against both supported OpenClaw versions.
- Tests cover both runtime capability shapes for durable receive and access-group command authorization.
- Outbound media remains typed without importing package-private files.
- The compatibility matrix and release documentation identify both tested host versions and any upstream 8.1 packaging defect.
- A protected live smoke is run only after the core 8.1 Gateway regressions are resolved; the current live Gateway remains on 7.1-2.

## Related work

- #27 depends on reliable durable completion before marking messages read.
- #68 should validate the trusted-install durable path after this compatibility work lands.

### PR DESCRIPTION
**Pull request by FtlC-ian**
Closes #71

## Summary

- defer live delivery when the legacy completion lookup fails after queue acceptance
- make legacy completion reconciliation durable by extending the tombstone to the queue pending lifetime
- delete stale queue-pending records when completion reconciliation fails
- retry only the affected durable message in process, with bounded exponential backoff
- serialize and coalesce deferred replay without replaying unrelated in-flight messages
- add regression coverage for every late review finding from #72 and #73

## Verification

- OpenClaw 2026.7.1-2: TypeScript build, 288 Vitest tests, 59 offline smoke tests
- OpenClaw 2026.8.1: TypeScript build, 288 Vitest tests, 59 offline smoke tests
- live Gateway unchanged on 2026.7.1-2

## Context

This is the corrective follow-up to #72. Hosted Codex surfaced two valid inline findings only after its summary switched to completed and the merge command had already started. Issue #71 was reopened immediately and remains open until this PR is reviewed and merged.

---

> [!NOTE]
> **Medium Risk**
> Changes message delivery timing and journal reconciliation on failure paths; incorrect deferral could delay or duplicate replies, but scope is limited to durable inbound edge cases with extensive test coverage.
> 
> **Overview**
> Hardens **Zulip durable inbound** handling when legacy tombstones and the queue journal disagree or storage calls fail transiently.
> 
> In **`durable-receive.ts`**, failed queue `complete` during legacy reconciliation now **extends the completed tombstone** to pending TTL and **drops stale queue-pending** rows when extension succeeds; if both complete and extension fail, the queue entry is left intact. **`accept`** treats a **failed post-accept legacy completed lookup** as still-pending with **`retryInProcess`** instead of delivering live, and wraps reconciliation lookups in swallowing try/catch so duplicates stay safe.
> 
> In **`monitor.ts`**, delivery returns **`settled` / `pending-delivery` / `pending-completion`**. A new **deferred replay loop** (bounded exponential backoff, serialized task, wake on schedule) retries only affected IDs: full replay after lookup deferrals, **completion-only** after the user-visible reply succeeded, without blocking other in-flight messages. **`retryInProcess`** from `accept` schedules that path without restart.
> 
> **`monitor.test.ts`** adds synthetic failure hooks and regression tests for tombstone extension, lookup deferral, pending-list errors, completion-only retries, starvation avoidance, and serialization.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit b7d63ead79b4e5075cef9de7135054ee28691221. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/zulip/durable-receive.ts [hunk 1]
```diff
@@ -134,18 +148,31 @@ export function createZulipDurableInboundReceiveJournal(accountId: string) {
...
         return { kind: "pending" as const, duplicate: true as const, record: pending };
       }
       const result = await queueJournal.accept(key, payload, options);
-      const completedAfterAccept = await legacyCompletedStore.lookup(key);
+      let completedAfterAccept: LegacyCompletedRecord | undefined;
+      try {
+        completedAfterAccept = await legacyCompletedStore.lookup(key);
+      } catch {
+        if (result.kind === "accepted") {
+          return { kind: "pending" as const, duplicate: true as const, record: result.record };
```
[the remark this answers is not in the document]
Comment (FtlC-ian) [reply src/zulip/durable-receive.ts]: Fixed in 00c4557: a post-accept lookup deferral now carries an explicit retry marker, and the monitor schedules one coalesced in-process durable replay immediately. The replay task is tracked through monitor shutdown. Regression coverage proves delivery resumes without restart; exact 7.1-2 and 8.1 matrices each pass 279 unit tests and 59 smoke tests.

**Comment by FtlC-ian** [comment]
@codex review

File: src/zulip/durable-receive.ts [hunk 2]
```diff
@@ -134,18 +148,31 @@ export function createZulipDurableInboundReceiveJournal(accountId: string) {
...
+        }
+        return result;
+      }
       if (!completedAfterAccept) {
         return result;
       }
-      await reconcileLegacyCompletion(key, completedAfterAccept);
+      try {
+        await reconcileLegacyCompletion(key, completedAfterAccept);
+      } catch {}
```
[the remark this answers is not in the document]
Comment (FtlC-ian) [reply src/zulip/durable-receive.ts:178]: Fixed in 9de9d95: pending reconciliation now suppresses the already-completed record and continues startup even when completion, tombstone extension, and pending deletion all fail. A regression test forces all three failures and proves pending() still resolves without replay.

File: src/zulip/monitor.ts [hunk 3]
```diff
@@ -1666,6 +1666,28 @@ export async function monitorZulipProvider(opts: MonitorZulipOpts = {}): Promise
...
 
+  let replayPendingDurableInboundMessages: () => Promise<void> = async () => {};
+  let pendingDurableReplayScheduled = false;
+  const schedulePendingDurableReplay = () => {
+    if (pendingDurableReplayScheduled) {
+      return;
+    }
+    pendingDurableReplayScheduled = true;
+    const replayTask = Promise.resolve().then(async () => {
+      pendingDurableReplayScheduled = false;
```
[the remark this answers is not in the document]
Comment (FtlC-ian) [reply src/zulip/monitor.ts]: Fixed in 9de9d95: deferred replay uses one running pass plus one coalesced queued follow-up. A two-message overlap test blocks the first replay while the second deferral arrives and proves no concurrent duplicate dispatch. Exact 7.1-2 and 8.1 matrices each pass 281 unit tests and 59 smoke tests.

**Comment by FtlC-ian** [comment]
@codex review

File: src/zulip/monitor.ts [hunk 4]
```diff
@@ -1666,6 +1666,28 @@ export async function monitorZulipProvider(opts: MonitorZulipOpts = {}): Promise
...
+      }
+      await replayPendingDurableInboundMessages();
+    }).catch((err) => {
+        runtime.error?.(`zulip: deferred durable replay failed: ${String(err)}`);
+    });
+    activeMessageTasks.add(replayTask);
+    void replayTask.finally(() => {
+      activeMessageTasks.delete(replayTask);
+    });
+  };
```
[the remark this answers is not in the document]
Comment (FtlC-ian) [reply src/zulip/monitor.ts:1793]: Fixed in 3a740e8. Deferred replay is now keyed to the exact durable ID, so another in-flight pending record is never redispatched. Added a two-message regression that asserts each Zulip message ID is dispatched exactly once.

File: src/zulip/monitor.ts [hunk 5]
```diff
@@ -1666,6 +1666,31 @@ export async function monitorZulipProvider(opts: MonitorZulipOpts = {}): Promise
...
+    pendingDurableReplayRunning = true;
+    const replayTask = Promise.resolve().then(async () => {
+      while (pendingDurableReplayQueued && !opts.abortSignal?.aborted) {
+        pendingDurableReplayQueued = false;
+        await replayPendingDurableInboundMessages();
+      }
+    }).catch((err) => {
+      runtime.error?.(`zulip: deferred durable replay failed: ${String(err)}`);
+    }).finally(() => {
+      pendingDurableReplayRunning = false;
```
[the remark this answers is not in the document]
Comment (FtlC-ian) [reply src/zulip/monitor.ts]: Fixed in 3a740e8. The worker owns a single exact task promise, drains the queued ID set, and atomically clears/restarts from its finally handler. The serialized two-message regression covers the follow-up scheduling boundary.

File: src/zulip/monitor.ts [hunk 6]
```diff
@@ -1666,6 +1666,31 @@ export async function monitorZulipProvider(opts: MonitorZulipOpts = {}): Promise
...
+      return;
+    }
+    pendingDurableReplayRunning = true;
+    const replayTask = Promise.resolve().then(async () => {
+      while (pendingDurableReplayQueued && !opts.abortSignal?.aborted) {
+        pendingDurableReplayQueued = false;
+        await replayPendingDurableInboundMessages();
+      }
+    }).catch((err) => {
+      runtime.error?.(`zulip: deferred durable replay failed: ${String(err)}`);
```
[the remark this answers is not in the document]
Comment (FtlC-ian) [reply src/zulip/monitor.ts]: Fixed in 3a740e8. Transient pending-list failures now retain the exact durable ID and retry with bounded exponential backoff from 250 ms to 30 seconds until success or monitor abort. Added a regression proving recovery without restart.

**Comment by FtlC-ian** [comment]
@codex review

File: src/zulip/durable-receive.ts [hunk 7]
```diff
@@ -134,18 +142,36 @@ export function createZulipDurableInboundReceiveJournal(accountId: string) {
...
+        if (result.kind === "accepted") {
+          return {
+            kind: "pending" as const,
+            duplicate: true as const,
+            record: result.record,
+            retryInProcess: true as const,
+          };
+        }
+        return result;
+      }
```
[the remark this answers is not in the document]
Comment (FtlC-ian) [reply src/zulip/durable-receive.ts:172]: Fixed in 3a740e8. The exact durable ID remains queued across transient pending-list failures and retries with bounded exponential backoff until success or abort. Worker ownership is atomically cleared and restarted in finally, closing the queued-follow-up boundary. Dedicated regressions cover both repeated read failure and overlapping two-message scheduling.

File: src/zulip/monitor.ts [hunk 8]
```diff
@@ -1666,6 +1666,57 @@ export async function monitorZulipProvider(opts: MonitorZulipOpts = {}): Promise
...
+          if (record) {
+            const payload = record.payload as ZulipDurableInboundPayload;
+            await deliverDurableInboundMessage(
+              record.id,
+              deserializeZulipDurableInboundMessage<ZulipMessage>(payload.message),
+              record.metadata,
+              { replay: true },
+            );
+          }
+          deferredDurableReplayIds.delete(durableId);
```
[the remark this answers is not in the document]
Comment (FtlC-ian) [reply src/zulip/monitor.ts:1775]: Fixed in 7ccad6c. Durable delivery now reports settled versus pending; deferred IDs are removed only after settlement. Retryable stream-policy, delivery, abort, and exception outcomes retain the ID and retry with bounded backoff. A regression forces one replay metadata failure, restores metadata, and proves same-process completion without restart.

**Comment by FtlC-ian** [comment]
@codex review

File: src/zulip/monitor.ts [hunk 9]
```diff
@@ -1663,9 +1664,66 @@ export async function monitorZulipProvider(opts: MonitorZulipOpts = {}): Promise
...
+            const disposition = await deliverDurableInboundMessage(
+              record.id,
+              deserializeZulipDurableInboundMessage<ZulipMessage>(payload.message),
+              record.metadata,
+              { replay: true },
+            );
+            if (disposition === "pending") {
+              await delay(retryDelayMs, opts.abortSignal);
+              retryDelayMs = Math.min(30000, retryDelayMs * 2);
+              continue;
```
[the remark this answers is not in the document]
Comment (FtlC-ian) [reply src/zulip/monitor.ts]: Fixed in ddeaea7. Deferred retry state is now per durable ID with independent next-attempt timestamps and bounded backoff. A retryable ID is rotated behind ready IDs, and the worker selects the next eligible record rather than blocking the queue. A regression keeps one stream record permanently retryable and proves a later DM still completes.

**Comment by FtlC-ian** [comment]
@codex review

File: src/zulip/durable-receive.ts [hunk 10]
```diff
@@ -120,8 +120,16 @@ export function createZulipDurableInboundReceiveJournal(accountId: string) {
...
       return;
+    } catch {
+      try {
+        await legacyCompletedStore.register(id, completed, {
+          ttlMs: ZULIP_DURABLE_INBOUND_PENDING_TTL_MS,
+        });
+      } catch {}
+      try {
+        await queueJournal.deletePending(id);
+      } catch {}
```
[the remark this answers is not in the document]
Comment (FtlC-ian) [reply src/zulip/durable-receive.ts]: Fixed in b7db535. Queue deletion now occurs only after the legacy completion tombstone is successfully extended to the queue pending lifetime. If extension fails, the queue-backed pending record remains available for later reconciliation. The all-writes-fail regression now asserts that delete is never attempted.

**Comment by FtlC-ian** [comment]
@codex review

File: src/zulip/monitor.ts [hunk 11]
```diff
@@ -1663,9 +1664,97 @@ export async function monitorZulipProvider(opts: MonitorZulipOpts = {}): Promise
...
+            const payload = record.payload as ZulipDurableInboundPayload;
+            const disposition = await deliverDurableInboundMessage(
+              record.id,
+              deserializeZulipDurableInboundMessage<ZulipMessage>(payload.message),
+              record.metadata,
+              { replay: true },
+            );
+            if (disposition === "pending") {
+              deferDurableReplayRetry(durableId);
+              continue;
```
[the remark this answers is not in the document]
Comment (FtlC-ian) [reply src/zulip/monitor.ts]: Fixed in 0d9a1dc. Successful delivery and journal completion are now separate states: a completion failure records a completion-only retry, and the worker calls the journal completion path directly without invoking message handling or reply dispatch again. The regression test forces the first queue completion to fail and proves two completion attempts but exactly one context build and one dispatch.

File: src/zulip/monitor.ts [hunk 12]
```diff
@@ -1663,9 +1664,97 @@ export async function monitorZulipProvider(opts: MonitorZulipOpts = {}): Promise
...
+            deferredDurableRetryState.get(candidateId)?.nextAttemptAt ?? 0;
+          if (nextAttemptAt <= now) {
+            durableId = candidateId;
+            break;
+          }
+          waitMs = Math.min(waitMs, nextAttemptAt - now);
+        }
+        if (!durableId) {
+          await delay(waitMs, opts.abortSignal);
+          continue;
```
[the remark this answers is not in the document]
Comment (FtlC-ian) [reply src/zulip/monitor.ts:1748]: Fixed in 0d9a1dc. The deferred worker's backoff wait is now explicitly wakeable when a newly accepted durable ID is scheduled, so a ready message does not inherit another ID's retry delay. The strengthened fairness regression queues a later ready DM during the first ID's backoff and proves the first ID is released only once before that DM dispatches.

**Comment by FtlC-ian** [comment]
@codex review

**Comment by FtlC-ian** [comment]
@codex review
